apps/src/lib/levelbuilder/code-docs-editor/ProgrammingEnvironmentEditor.jsx

```diff
@@ -0,0 +1,164 @@
+
+const EDITOR_TYPES = ['blockly', 'droplet', 'text'];
+
+const useProgrammingEnvironment = initialProgrammingEnvironment => {
+  const [programmingEnvironment, setProgrammingEnvironment] = useState(
+    initialProgrammingEnvironment
+  );
+  const updateProgrammingEnvironment = (key, value) => {
+    setProgrammingEnvironment({...programmingEnvironment, [key]: value});
+  };
+
+  return [programmingEnvironment, updateProgrammingEnvironment];
+};
+
+export default function ProgrammingEnvironmentEditor({
+  initialProgrammingEnvironment
+}) {
+  const {
+    name,
+    ...remainingProgrammingEnvironment
+  } = initialProgrammingEnvironment;
+  const [
+    programmingEnvironment,
+    updateProgrammingEnvironment
+  ] = useProgrammingEnvironment(remainingProgrammingEnvironment);
+  const [uploadImageDialogOpen, setUploadImageDialogOpen] = useState(false);
+  const [isSaving, setIsSaving] = useState(false);
+  const [lastUpdated, setLastUpdated] = useState(null);
+  const [error, setError] = useState(null);
+
+  const save = () => {
+    if (isSaving) {
+      return;
+    }
+    setIsSaving(true);
+    fetch(`/programming_environments/${name}`, {
+      method: 'PUT',
+      headers: {
+        'content-type': 'application/json',
+        'X-CSRF-Token': $('meta[name="csrf-token"]').attr('content')
+      },
+      body: JSON.stringify(programmingEnvironment)
+    })
+      .then(response => {
+        setIsSaving(false);
+        if (response.ok) {
+          setLastUpdated(Date.now());
+          setError(null);
+        } else {
+          setError(response.statusText);
+        }
+      })
+      .catch(error => {
+        setIsSaving(false);
+        setError(error.responseText);
+      });
+  };
+
+  return (
+    <div>
+      <h1>{`Editing ${name}`}</h1>
+      <h2>
+        This feature is in development. Please continue to use curriculum
+        builder to edit code documentation.
+      </h2>
+      <label>
+        Title
+        <input
+          value={programmingEnvironment.title || ''}
+          onChange={e => updateProgrammingEnvironment('title', e.target.value)}
+          style={styles.textInput}
+        />
+      </label>
+      <label>
+        IDE URL (Slug)
+        <input value={name} style={styles.textInput} readOnly />
+      </label>
+      <label>
+        Image
+        <Button
+          onClick={() => setUploadImageDialogOpen(true)}
+          text="Choose Image"
+          color="gray"
+          icon="plus-circle"
+        />
+        {programmingEnvironment.imageUrl && (
+          <span>{programmingEnvironment.imageUrl}</span>
+        )}
+      </label>
+
+      <TextareaWithMarkdownPreview
+        markdown={programmingEnvironment.description || ''}
+        label={'Description'}
+        handleMarkdownChange={e =>
+          updateProgrammingEnvironment('description', e.target.value)
+        }
+        features={{imageUpload: true}}
+      />
+      <label>
+        How should this document render?
```

Comment:
UI Thoughts: I find this dropdown kinda hard to see at its current location because its below the big description area.

Reply:
Maybe move it up with the image selection?

Reply:
It does look a bit better with the Categories section added (WIP) but I think it's still a bit easy to miss. Plus it's pretty important, so I'll move it up.
<img width="1076" alt="Screen Shot 2022-01-05 at 10 18 32 AM" src="https://user-images.githubusercontent.com/46464143/148268335-ee170c8a-c932-49a1-b90a-f45ce222dc9f.png">